proposals/pubsub-mq-interfaces-proposal.md

```diff
@@ -0,0 +1,144 @@
+
+###  2.3. <a name='pubsubmessage-queueComparison'></a>`pubsub`/`message-queue` Comparison
+
+Historically, `pubsub` services (i.e., implementers of the `pubsub` pattern)  are thought to be comprised of non-competitive consumers – That is, if a message is sent to a channel, every receiver that is subscribed to that channel will, eventually, receive it.
+
+While, on the other hand, historically, `message-queue`s are thought to be comprised of competitive consumers – That is, a queued message is dequeued as soon as any one receiver has read it.
+
+However, if we look at services that provide `message-queue` / `pubsub` implementations, the line between the two concepts becomes blurred.
+
+That is, today's most used related services provide both `message-queue`, and `pubsub` capabilities. Below, is a visual example from [HiveMQ](https://www.hivemq.com/) displaying an event-driven architecture with both `message-queue`, and `pubsub` capabilities:
+
+![](./images/pubsub-interface-proposal-0.gif)
+Source: [HiveMQ's Shared Subscriptions](https://www.hivemq.com/docs/hivemq/4.9/user-guide/shared-subscriptions.html)
+
+##  3. <a name='pubsubhttpInteroptability'></a>`pubsub`/`http` Interoptability
+
+`pubsub`s are often used in conjuction with `http` to send, and receive messages from subscribed topics. However, through experimentation on the intereoperability between the two services, it is impossible to use `slight`'s `http` service to correctly manage `pubsub` subscriptions. This is because the `http` interface cannot return a handle to the `pubsub` client under which the subscription was created - Hence, you can only receive messages from a subscribed topic if the subscriptions is done in conjunction with the message polling. This is a problem because message polling, and subscribing should be two separate concerns as, if you only subscribe to a topic at the time of polling, you will miss messages that were sent to the topic before you subscribed to it.
+
+##  4. <a name='ProblemStatement'></a>Problem Statement
+
+1. As one of SpiderLightning's core goals is to stray away from mirroring infrastructure and focus on concepts, we should unify the `pubsub`, and `message-queue` interfaces into the umbrella of the `pubsub` pattern.
+1. The current `pubsub` interface is not well suited for event-driven architectures as:
+	- the `payload` type is lacking, and
+	- `http` interoptability is lacking - It should be modified to allow better subscription management.
+
+##  5. <a name='Proposal'></a>Proposal
+
+The change involves:
+- unify `pub` and `sub` resources into a single `pubsub` resource,
+- modify the `pubsub` interface to contain `message-queue` semantics,
+- change the `payload` type to `event` (following the `CloudEvents` spec),
+- remove the `subscribe` method from the `pubsub` interface to pass the responsibility of subscriptions to `slight` through information made available at compilation-time through the `slightfile`, and
```

Comment:
Probably should explain why we favor a static subscription model instead of dynamic subscription model. That is, we favor users to define what topics/queues to subscribe to in a static file, instead of in code / runtime.

Reply:
done ✅